assigner/backends/base.py

```diff
@@ -0,0 +1,154 @@
+import git
+import re
+from typing import Optional
+
+
+class RepoError(Exception):
+    pass
+
+
+class RepoBase(object):
+    """Generic Repo base"""
+
+    PATH_RE = re.compile(
+        r"^/(?P<namespace>[\w\-\.]+)/(?P<name>[\w\-\.]+)\.git$"
+    )
+
+    def __init__(self, url_base: str, namespace: str, name: str, token: str, url: Optional[str] = None):
+        raise NotImplementedError
+
+    @classmethod
+    def build_url(cls, url_base: str, namespace: str, name: str) -> str:
+        """ Builds a url for a repository """
+        raise NotImplementedError
+
+    @classmethod
+    def from_url(cls, url: str, token: str):
```

Comment:
I looked into returning a self type from a class method in type annotations. 

I think this is the recommended way:
https://github.com/python/mypy/issues/1212#issuecomment-250526301

Though I also saw this:
https://github.com/python/mypy/issues/1212#issuecomment-219253128

Thoughts? I really like these type annotations; I'll keep adding them where I can.

Reply:
I think https://github.com/python/peps/commit/ada7d3566e26edf5381d1339b61e48a82c51c566 is what you want?

I also am a big fan of the type annotations; eventually I want to be able to integrate mypy here but there's a bit of footwork involved in writing annotations for some of the libraries we depend on.

Reply:
Yeah, I think that's what I want. I tried reading this at 7 last night and I couldn't comprehend it.